UI/MemStickScreen.cpp

```diff
@@ -0,0 +1,155 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License 2.0 for more details.
+
+// A copy of the GPL 2.0 should have been included with the program.
+// If not, see http://www.gnu.org/licenses/
+
+// Official git repository and contact information can be found at
+// https://github.com/hrydgard/ppsspp and http://www.ppsspp.org/.
+
+#include "ppsspp_config.h"
+
+#include "android/jni/app-android.h"
+
+#include "Common/Log.h"
+#include "Common/UI/UI.h"
+#include "Common/UI/View.h"
+#include "Common/UI/ViewGroup.h"
+
+#include "Common/StringUtils.h"
+#include "Common/File/FileUtil.h"
+#include "Common/File/Path.h"
+#include "Common/System/System.h"
+#include "Common/System/NativeApp.h"
+#include "Common/Data/Text/I18n.h"
+#include "Common/System/Display.h"
+
+#include "Core/Util/GameManager.h"
+#include "Core/System.h"
+
+#include "UI/MemStickScreen.h"
+#include "UI/MainScreen.h"
+#include "UI/MiscScreens.h"
+
+#include "Core/Config.h"
+
+MemStickScreen::~MemStickScreen() { }
+
+void MemStickScreen::CreateViews() {
+	using namespace UI;
+
+	auto di = GetI18NCategory("Dialog");
+	auto iz = GetI18NCategory("MemStick");
+
+	Margins actionMenuMargins(15, 15, 15, 0);
+
+	root_ = new AnchorLayout();
+
+	ViewGroup *columns = new LinearLayout(ORIENT_HORIZONTAL, new AnchorLayoutParams(FILL_PARENT, FILL_PARENT));
+	root_->Add(columns);
+
+	ViewGroup *leftColumn = new AnchorLayout(new LinearLayoutParams(1.0));
+	ViewGroup *rightColumnItems = new LinearLayout(ORIENT_VERTICAL, new LinearLayoutParams(300, FILL_PARENT, actionMenuMargins));
+	columns->Add(leftColumn);
+	columns->Add(rightColumnItems);
+
+	Path path = g_Config.memStickDirectory;
+	int64_t freeSpaceAtMemStick = -1;
+#if PPSSPP_PLATFORM(ANDROID)
+	if (Android_IsContentUri(path.ToString())) {
+		freeSpaceAtMemStick = Android_GetFreeSpaceByContentUri(path.ToString());
+	} else {
+		freeSpaceAtMemStick = Android_GetFreeSpaceByFilePath(path.ToString());
+	}
+#endif
+
+	int leftSide = 100;
+	settingInfo_ = new SettingInfoMessage(ALIGN_CENTER | FLAG_WRAP_TEXT, new AnchorLayoutParams(dp_xres - leftSide - 40.0f, WRAP_CONTENT, leftSide, dp_yres - 80.0f - 40.0f, NONE, NONE));
+	settingInfo_->SetBottomCutoff(dp_yres - 200.0f);
+
+	root_->Add(settingInfo_);
+
+	leftColumn->Add(new TextView(iz->T("Memory Stick Storage"), ALIGN_LEFT, false, new AnchorLayoutParams(10, 10, NONE, NONE)));
+	leftColumn->Add(new TextView(iz->T("MemoryStickDescription", "Choose where your PSP memory stick data (savegames, etc) is stored"), ALIGN_LEFT, false, new AnchorLayoutParams(10, 50, NONE, NONE)));
+	leftColumn->Add(new TextView(g_Config.memStickDirectory.ToVisualString(), ALIGN_LEFT, false, new AnchorLayoutParams(10, 140, NONE, NONE)));
+
+	std::string freeSpaceText = "Free space: N/A";
+	if (freeSpaceAtMemStick >= 0) {
+		freeSpaceText = StringFromFormat("free space: %lld bytes", freeSpaceAtMemStick);
```

Comment:
Hm.  Maybe we should at least say MB if it's >= 2048 * 1024?  Definitely nice to include free space.

-[Unknown]

Reply:
Lots to improve in this UI, but yeah, that one I'll do right away.